runtime/vm/resolvesupport.cpp

```diff
@@ -872,8 +872,32 @@ resolveInstanceFieldRefInto(J9VMThread *vmStruct, J9Method *method, J9ConstantPo
 				targetClass = currentTargetClass;
 				goto illegalAccess;
 			}
+			
+			/* TODO: Redo value class check when L-World spec is finalized */
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+			/* When setting a field, throw IncompatibleClassChangeError if the flag
+			 * J9_RESOLVE_FLAG_CHECK_VALUE_CLASS is set but resolvedClass is not a value type
+			 * or if J9_RESOLVE_FLAG_CHECK_VALUE_CLASS is not set and the resolved class is a
+			 * value type.
+			 */
+			if (J9_ARE_ALL_BITS_SET(resolveFlags, J9_RESOLVE_FLAG_FIELD_SETTER)
+				&& (J9_ARE_ALL_BITS_SET(resolveFlags, J9_RESOLVE_FLAG_CHECK_VALUE_CLASS)
+					!= J9_ARE_ALL_BITS_SET(resolvedClass->romClass->modifiers, J9AccValueType))
+			) {
+					setCurrentException(vmStruct, J9VMCONSTANTPOOL_JAVALANGINCOMPATIBLECLASSCHANGEERROR, NULL);
+					goto done;
+			}
 
-			if ((resolveFlags & J9_RESOLVE_FLAG_FIELD_SETTER) != 0 && (modifiers & J9_JAVA_FINAL) != 0) {
+			/* Field must be final when class is a value class */
+			if (J9_ARE_ALL_BITS_SET(resolveFlags, J9_RESOLVE_FLAG_FIELD_SETTER)
+				&& (J9_ARE_ALL_BITS_SET(modifiers, J9_JAVA_FINAL)
+				|| J9_ARE_ALL_BITS_SET(resolveFlags, J9_RESOLVE_FLAG_CHECK_VALUE_CLASS))
```

Comment:
Is it just the or clause that is `J9VM_OPT_VALHALLA_VALUE_TYPES` specific?  Can just that clause be under ifdef?